slang.h

```diff
@@ -810,6 +829,51 @@ extern "C"
 
     #define SLANG_UUID_ISlangFileSystemExt { 0x5fb632d2, 0x979d, 0x4481, { 0x9f, 0xee, 0x66, 0x3c, 0x3f, 0x14, 0x49, 0xe1 } }
 
+    /* Identifies different types of writer target*/
+    typedef unsigned int SlangWriterChannel;
+    enum
+    {
+        SLANG_WRITER_CHANNEL_DIAGNOSTIC,
+        SLANG_WRITER_CHANNEL_STD_OUTPUT,
+        SLANG_WRITER_CHANNEL_STD_ERROR,
+        SLANG_WRITER_CHANNEL_COUNT_OF,
+    };
+
+    typedef unsigned int SlangWriterMode;
+    enum 
+    {
+        SLANG_WRITER_MODE_TEXT,
+        SLANG_WRITER_MODE_BINARY,
+    };
+
+    /** A stream typically of text, used for outputting diagnostic as well as other information.
+    */
+    struct ISlangWriter : public ISlangUnknown
+    {
+    public:
+            /** Write the formatted string with the used args. If returns SLANG_E_NOT_IMPLEMENTED, will use a default internal implementation, and call write with result
+            @param format Format string
+            @param args The arguments
+            @return SLANG_E_NOT_IMPLEMENTED if not implemented, SLANG_OK on success */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL writeVaList(const char* format, va_list args) = 0;
```

Comment:
I did this somewhat reluctantly, not least because valist isn't something you can have strictly in a ABI interface if you want to work across compilers. Passing in valist meant that I can write formatted strings directly to a stream. 
To avoid this I could write the formatted string and just use write. The problem there is that for every formatted write, I would likely need an allocation/free (could be limited with a stack based fixed size buffer). 
To work around this I have a PR which ISlangWriter has an associated 'appendBuffer', that avoids a copy if the implementation just needs to write to memory, and avoids most allocs. 
This means for a FILE stream, the formatted strings are converted into writes using that buffer. This may be slightly less efficient, because we first need to calculate the buffer size, by parsing the format string and parameters, then allocate then create the format string and finally write.